homeassistant/components/http/static.py

```diff
@@ -9,28 +9,51 @@
 from aiohttp.web import FileResponse, Request, StreamResponse
 from aiohttp.web_exceptions import HTTPForbidden, HTTPNotFound
 from aiohttp.web_urldispatcher import StaticResource
+from lru import LRU  # pylint: disable=no-name-in-module
+
+from homeassistant.core import HomeAssistant
+
+from .const import KEY_HASS
 
 CACHE_TIME: Final = 31 * 86400  # = 1 month
 CACHE_HEADERS: Final[Mapping[str, str]] = {
     hdrs.CACHE_CONTROL: f"public, max-age={CACHE_TIME}"
 }
+PATH_CACHE = LRU(512)
+
+
+def _get_file_path(
+    filename: str, directory: Path, follow_symlinks: bool
+) -> Path | None:
+    filepath = directory.joinpath(filename).resolve()
+    if not follow_symlinks:
+        filepath.relative_to(directory)
+    # on opening a dir, load its contents if allowed
+    if filepath.is_dir():
+        return None
+    if filepath.is_file():
+        return filepath
+    raise HTTPNotFound
 
 
 class CachingStaticResource(StaticResource):
     """Static Resource handler that will add cache headers."""
 
     async def _handle(self, request: Request) -> StreamResponse:
         rel_url = request.match_info["filename"]
+        hass: HomeAssistant = request.app[KEY_HASS]
+        filename = Path(rel_url)
+        if filename.anchor:
```

Comment:
I can't get this to actually hit in a test, but its copied from aiohttp so best to let it be even if we can't cover it